[Modules][Diagnostic] Mention which AST file's options differ from the current TU options.

vsapsai opened the pull request:
Claiming a mismatch is always in a precompiled header is wrong and misleading as a mismatch can happen in any provided AST file. Emitting a path for a file with a problem allows to disambiguate between multiple input files.

Use generic term "AST file" because we don't always know a kind of the provided file (for example, see `ASTReader::readASTFileControlBlock`).

rdar://65005546

vsapsai committed:
[Modules][Diagnostic] Mention which AST file's options differ from the current TU options.

Claiming a mismatch is always in a precompiled header is wrong and
misleading as a mismatch can happen in any provided AST file. Emitting
a path for a file with a problem allows to disambiguate between multiple
input files.

Use generic term "AST file" because we don't always know a kind of
the provided file (for example, see `ASTReader::readASTFileControlBlock`).

rdar://65005546

vsapsai commented:
Sorry that the change is pretty big. But it is fairly mechanical. If you have ideas about simplifying the review process, I'm open to that. Right now I'm thinking if I should make the logical changes first, you review them, and I commit clang-formatted change.

jansvoboda11 commented:
> Sorry that the change is pretty big. But it is fairly mechanical. If you have ideas about simplifying the review process, I'm open to that. Right now I'm thinking if I should make the logical changes first, you review them, and I commit clang-formatted change.

I'd suggest keeping this PR and the resulting commit free of any formatting changes and then commit the results from `clang-format` as a separate commit.

After https://github.com/llvm/llvm-project/pull/101280 I had the impression that we're trying to be very precise about reporting whether an AST file is a precompiled header, module file, or other. Why doesn't this PR implement that? To be clear this change is already a big improvement as is, but I'd like to understand if further polish and unification is still planned.

ChuanqiXu9 reviewed (approved):
I did a quick scanning and it looks good except formatting. Please leave a chance to @jansvanboda to an another look.

vsapsai committed:
Add a missing file with changes.

vsapsai commented:
> I'd suggest keeping this PR and the resulting commit free of any formatting changes and then commit the results from `clang-format` as a separate commit.

Will do that.

> After #101280 I had the impression that we're trying to be very precise about reporting whether an AST file is a precompiled header, module file, or other. Why doesn't this PR implement that? To be clear this change is already a big improvement as is, but I'd like to understand if further polish and unification is still planned.

I was precise in #101280 because it was easy to do, not because I was trying hard. The main reason I didn't use a similar approach in this PR is because ModuleKind isn't always available. For example, the diagnostic can be reached through a call stack like

```
ASTReader::isAcceptableASTFile
  ASTReader::readASTFileControlBlock
    ASTReader::ReadOptionsBlock
      ASTReader::ParseLanguageOptions
        ASTReaderListener::ReadLanguageOptions
          checkLanguageOptions          
```

and my understanding is that `isAcceptableASTFile` doesn't know the module kind by design. There are ways to deal with it, of course. I had in mind passing `std::optional<ModuleKind>` alongside `StringRef Filename` everywhere but don't think the extra complexity is worth the resulting precision.

As far as I remember, module kind is decided during a module file creation and we don't support interpreting the same .pcm file as a different kind (cannot use PCH as a preamble, for instance). That's why I think emitting a module kind in a diagnostic isn't particularly valuable. Over time that can change but right now I'm not planning a further polish.

jansvoboda11 reviewed (approved):
Makes sense, thank you!

vsapsai committed:
Undo formatting changes to simplify reviewing.

vsapsai committed:
Tweak naming - use ModuleFilename instead of Filename in more places.

For ASTListener Filename is too generic, not entirely clear what file
is implied here, so ModuleFilename should help with that. No changes for
ASTReader as I believe it is clear that in `ASTReader::ReadOptionsBlock`
Filename means the name of the file being read.

vsapsai commented on clang/include/clang/Serialization/ASTReader.h:
```diff
@@ -130,7 +130,7 @@ class ASTReaderListener {
   /// \returns true to indicate the options are invalid or false otherwise.
   virtual bool ReadLanguageOptions(const LangOptions &LangOpts,
-                                   bool Complain,
+                                   StringRef Filename, bool Complain,
```
Tweaked the name. If there are no further comments, I'll merge the PR tomorrow and follow-up with a formatting change.

vsapsai committed:
Missed 1 Filename to ModuleFilename rename.

vsapsai commented:
Formatting change was #102484.
